asf_heat_pump_suitability/analysis/hn_zones/comparison_of_hn_zones.py

```diff
@@ -0,0 +1,350 @@
+        raise ValueError("pilot_score must be either 1 or 0")
+
+    # Filter and calculate the average score
+    filtered_la_hp_suitability_scores = la_hp_suitability_scores.filter(
+        pl.col("DESNZ_pilot_score") == pilot_score
+    )
+    avg_hn_score_pilot = (
+        filtered_la_hp_suitability_scores.select(
+            pl.col("HN_N_avg_score_weighted").mean()
+        ).item()
+        if filtered_la_hp_suitability_scores.height > 0
+        else 0.0
+    )
+    return avg_hn_score_pilot
+
+
+def calculate_mae_for_pilot_score(
+    hp_suitability_scores_with_desnz: pl.DataFrame, score_value: int
+) -> float:
+    """
+    Calculate and log the Mean Absolute Error (MAE) for entries with a specified 'DESNZ_pilot_score'.
+
+    Args:
+        hp_suitability_scores_with_desnz (pl.DataFrame): DataFrame containing 'DESNZ_pilot_score' and 'absolute_error' columns.
+        score_value (int): The value of 'DESNZ_pilot_score' (must be 1 or 0) for which to calculate the MAE.
+
+    Returns:
+        float: The MAE for the specified 'DESNZ_pilot_score'.
+
+    Raises:
+        ValueError: If score_value is not 1 or 0.
+    """
+    # Validate that score_value is either 1 or 0
+    if score_value not in {0, 1}:
+        raise ValueError("score_value must be either 1 or 0")
+
+    mae = hp_suitability_scores_with_desnz.filter(
+        pl.col("DESNZ_pilot_score") == score_value
+    )["absolute_error"].mean()
+
+    logging.info(
+        f"Mean Absolute Error (MAE) for DESNZ_pilot_score = {score_value}: {mae}"
+    )
+    return mae
+
+
+def calculate_mae_for_all(
+    hp_suitability_scores: pl.DataFrame, desnz_col: str, nesta_hn_score_col: str
+) -> Tuple[pl.DataFrame, float]:
+    """
+    Calculate the Mean Absolute Error (MAE) between the DESNZ pilot score and Nesta's HN suitability score, add the absolute error column to the DataFrame,
+    and log the result.
+
+    Args:
+        hp_suitability_scores_with_desnz (pl.DataFrame): DataFrame containing the DESNZ pilot (actual) and Nesta HN suitability score (predicted) columns.
+        desnz_col (str): Name of the column with DESNZ pilot (actual) values.
+        nesta_hn_score_col (str): Name of the column with Nesta HN suitability (predicted) values.
+
+    Returns:
+        Tuple[pl.DataFrame, float]:
+            - Updated DataFrame with 'absolute_error' column added.
+            - The Mean Absolute Error (MAE) between the actual and predicted columns.
+    """
+    # Calculate absolute error and add it as a new column
+    hp_suitability_scores_with_error = hp_suitability_scores.with_columns(
+        (pl.col(desnz_col) - pl.col(nesta_hn_score_col)).abs().alias("absolute_error")
+    )
+
+    # Calculate the mean of the absolute error
+    mae = hp_suitability_scores_with_error["absolute_error"].mean()
+    logging.info(
+        f"Mean Absolute Error (MAE) for {desnz_col} vs {nesta_hn_score_col}: {mae}"
+    )
+
+    return hp_suitability_scores_with_error, mae
+
+
+def main():
```

Comment:
I was using `main()` and `run()` functions previously. I like how clean they are but the downside is they don't work well with `python -i` when debugging or experiencing other issues. This is because if the `main()` function is interrupted then there are no variables saved to memory to investigate.

You could just replace the `def main():` line with `if __name__ == "__main__"` if you want to remove it.

Reply:
Ah ok, I haven't gotten into the habit of using `python -i` yet so didn't run into this as an issue. I've now changed it, likely I'll have to define these lines of code as a function (with a different name to main) anyway when scaling.